pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -245,9 +305,247 @@ func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
+		return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+	}
+
+	var err error = nil
+
+	if leaseStatus == LeaseStatusNotFound {
+		reqLogger.Infof("MaintOn: create new lease - no existing lase found")
+
+		if lease, err = r.createOrUpdateLease(node, lease, 0, TransitionSet, true, reqLogger); err != nil {
+			return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+		}
+	}
+
+	if leaseStatus == LeaseStatusOwnedByDifferentHolder {
+		reqLogger.Infof("MaintOn: update lease -  different owner & invalid lease ")
+
+		if _, err = r.createOrUpdateLease(node, lease, durationInSeconds, TransitionInc, false, reqLogger); err != nil {
+			return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+		}
+	}
+
+	if !isLeasePeriodSpecified(lease) || isLeaseDurationZero(lease) {
+		reqLogger.Infof("MaintOn: update lease - null lease period or zero duration")
+
+		if _, err = r.createOrUpdateLease(node, lease, durationInSeconds, TransitionInc, false, reqLogger); err != nil {
+			return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+		}
+
+	}
+
+	if r.isLeaseDurationChanged(lease, durationInSeconds) {
+		reqLogger.Info("MaintOn: update lease - lease duration change requested")
+
+		if _, err = r.createOrUpdateLease(node, lease, durationInSeconds, TransitionNoChange, false, reqLogger); err != nil {
+			return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+		}
+
+	}
+
+	if !isLeaseValid(lease, false) {
+		reqLogger.Info("MaintOn: update lease - lease expired")
+
+		if _, err = r.createOrUpdateLease(node, lease, durationInSeconds, TransitionNoChange, false, reqLogger); err != nil {
+			return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+		}
+	}
+	return true, reconcile.Result{}
+}
+
+func (r *ReconcileNodeMaintenance) doesLeaseExist(nodeName string, reqLogger *log.Entry) (LeaseStatus, error, *coordv1beta1.Lease) {
+	lease := &coordv1beta1.Lease{}
+
+	nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+
+	if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Infof("Lease object not found name=%s ns=%s", nName.Name, nName.Namespace)
+			return LeaseStatusNotFound, nil, nil
+		}
+
+		reqLogger.Errorf("failed to get lease object. name=%s ns=%s", nName.Name, nName.Namespace)
+		return LeaseStatusFail, err, nil
+	}
+
+	heldByMe := lease.Spec.HolderIdentity != nil && LeaseHolderIdentity == *lease.Spec.HolderIdentity
+	reqLogger.Debugf("got lease name=%s ns=%s heldByMe=%t", nName.Name, nName.Namespace, heldByMe)
+
+	if heldByMe {
+		return LeaseStatusOwnedByMe, nil, lease
+	}
+	return LeaseStatusOwnedByDifferentHolder, nil, lease
+}
+
+func (r *ReconcileNodeMaintenance) createOrUpdateLease(node *corev1.Node, lease *coordv1beta1.Lease, durationInSeconds int, transitions TransitionAction, resetDuration bool, reqLogger *log.Entry) (*coordv1beta1.Lease, error) {
```

Comment:
Your logic will flow better if you start with `getOrCreateLease()` (which creates an inactive lease if not found) and `updateLease()`

Reply:
Currently the lease period is a constant value, but you may want to change this in a possible future change the CRD, therefore I am passing these values a as arguments.

Reply:
done